harness/determined/pytorch/trainer.py

```diff
@@ -0,0 +1,285 @@
+import contextlib
+import warnings
+from enum import Enum
+
+import determined as det
+from determined import core, horovod, util
+from determined import pytorch
+from typing import Any, Callable, Dict, Iterator, List, Optional
+import torch
+import torch.nn as nn
+from determined.common import check
+from determined.horovod import hvd
+import torch.distributed as dist
+import logging
+
+from determined.pytorch import PyTorchTrialContext
+
+
+class TrainStepType(Enum):
+    EPOCH = "EPOCH"
+    BATCH = "BATCH"
+
+
+class Trainer:
+    def __init__(self, trial: pytorch.PyTorchTrial, context: PyTorchTrialContext):
+        self.trial = trial
+        self.context = context
+        self.core_context = self.context._core
+        self.train_loader = None
+        self.distributed_backend = det._DistributedBackend()
+        self.is_chief = self.core_context.distributed.rank == 0
+        self.callbacks = self.trial.build_callbacks()
+        for callback in self.callbacks.values():
+            if util.is_overridden(callback.on_checkpoint_end, pytorch.PyTorchCallback):
+                warnings.warn(
+                    "The on_checkpoint_end callback is deprecated, please use "
+                    "on_checkpoint_write_end instead",
+                    FutureWarning,
+                )
+
+    def _build_training_loader(self):
+        train_data = self.trial.build_training_data_loader()
+        num_replicas = self.core_context.distributed.size
+        rank = self.core_context.distributed.rank
+
+        training_loader = train_data.get_data_loader(
+            repeat=False, num_replicas=num_replicas, rank=rank
+        )
+        return training_loader
+
+    def _build_validation_loader(self):
+        val_data = self.trial.build_validation_data_loader()
+        if self.is_chief:
+            return val_data.get_data_loader(
+                repeat=False, skip=0, num_replicas=1, rank=0
+            )
+
+        num_replicas = self.core_context.distributed.size
+        rank = self.core_context.distributed.rank
+
+        val_loader = val_data.get_data_loader(
+            repeat=False, num_replicas=num_replicas, rank=rank
+        )
+        return val_loader
+
+    def train(
+        self,
+        max_epochs: Optional[int] = None,
+        # OR
+        max_batches: Optional[int] = None,
```

Comment:
why not something like:
`max_length=Epochs(10)` or `max_length=Batches(10000)`?